wave_optimization/include/wave/optimization/factor_graph/factor.hpp

```diff
@@ -0,0 +1,69 @@
+#ifndef WAVE_OPTIMIZATION_FACTOR_GRAPH_FACTOR_HPP
+#define WAVE_OPTIMIZATION_FACTOR_GRAPH_FACTOR_HPP
+
+#include <memory>
+
+#include "wave/utils/utils.hpp"
+#include "wave/optimization/factor_graph/variable.hpp"
+
+namespace wave {
+
+struct Factor {
+    const int arity;
+    const std::vector<std::shared_ptr<Variable>> variables;
+
+    Factor() : arity{0}, variables{} {}
+
+    Factor(const std::shared_ptr<Variable> &at) : arity{1}, variables{at} {}
+
+    Factor(const std::shared_ptr<Variable> &from,
+           const std::shared_ptr<Variable> &to)
+        : arity{2}, variables{from, to} {}
+
+    virtual std::string toString() {
+        std::ostringstream oss;
+
+        switch (this->arity) {
+            case 1:
+                oss << "[factor_type: unary, ";
+                oss << "at: " << this->variables[0]->toString();
+                oss << "]";
+                break;
+            case 2:
+                oss << "[factor_type: binary, ";
```

Comment:
A bit confusing - if the base class already knows everything about arity, what's the point of having separate `UnaryFactor` and `BinaryFactor` classes? E.g., why is this method `virtual`, but has this conditional inside it instead of being overridden in each derived class?